### PR TITLE
Initial version of slot-all element,

### PR DESCRIPTION
**Pull request by tomalec**
required for https://github.com/Starcounter/RebelsLounge/issues/368

### CODE REVIEW

**Comment by tomalec** [comment]
@platosha, would you mind reviewing?

Do you know a way to workaround polyfill bug in tests?
`querySelector('foo > slot')` does not work in polyfilled browsers, as it seems compound selectors with slots does not match.

**Comment by platosha** [comment]
@tomalec Indeed seems to be a bug in ShadyDOM.

The `children` list of the `#scope` element seems to be correct though, so I would suggest manual filtering of `children` as a workaround, for example:

```javascript
        it('should create `slot` elements for all slot names used by host\'s children, and `all-slotted` attribute to them', function() {
            const createdSlots = Array.from(host.shadowRoot.firstElementChild.children)
                .filter(el => el.matches('slot[name][all-slotted]'));
            expect(createdSlots).to.be.not.null;
            expect(createdSlots).to.have.length(2);
            Array.from(host.children).forEach((child)=>{
                child.hasAttribute('slot') &&
                    expect(createdSlots.find(el =>
                        el.name === child.getAttribute('slot')
                    )).to.be.not.null;
            });
        });
```

File: test/spec.html [hunk 1]
```diff
@@ -8,28 +8,146 @@
...
 
         it('should be upgraded element', function() {
-            expect(customElements.get('juicy-element')).to.be.defined;
-            expect(juicyElement).to.be.an.instanceOf(customElements.get('juicy-element'));
+            expect(customElements.get('slot-all')).to.be.defined;
+            expect(host.shadowRoot.querySelector('slot-all')).to.be.an.instanceOf(customElements.get('slot-all'));
+        });
+
+        it('should create `slot` elements for all slot names used by host\'s children, and `all-slotted` attribute to them', function() {
+            debugger
```
Comment (platosha) [inline test/spec.html]: Don’t forget to remove this `debugger`.

File: test/spec.html [hunk 2]
```diff
@@ -8,28 +8,146 @@
...
+            expect(host.shadowRoot.querySelector(`#scope > slot[name][all-slotted]`)).to.be.not.null;
+            expect(host.shadowRoot.querySelectorAll(`#scope > slot[name][all-slotted]`)).to.have.length(2);
+            Array.from(host.children).forEach((child)=>{
+                child.hasAttribute('slot') &&
+                expect(host.shadowRoot.querySelector(`#scope > slot[name="${child.getAttribute('slot')}"][all-slotted]`)).to.be.not.null;
```
Comment (platosha) [inline test/spec.html]: Hint: another option would be to check `child.assignedSlot`. Not a blocker though.

File: index.html [hunk 3]
```diff
@@ -1,21 +1,114 @@
...
     <meta charset="utf-8">
-    <title>&lt;juicy-element&gt;</title>
+    <title>&lt;slot-all&gt;</title>
     <link rel="stylesheet" href="http://juicy.github.io/github-markdown-css/github-markdown.css">
 
     <!-- Imports polyfill -->
     <script src="../webcomponentsjs/webcomponents-lite.js"></script>
 
+    <!-- Imports ace editor for nicer demo -->
+    <link rel="import" href="../juicy-ace-editor/juicy-ace-editor.html">
```
Comment (platosha) [inline index.html:10]: I’ve got 404 on this import. Should probably be added to `bower.json` as a dev dependency.

**Review by platosha — APPROVED** [review]
LGTM overall aside from test failures due to the ShadyDOM bug.

**Comment by platosha** [comment]
What about the wildcard names `<slot-all name="myapp/*"></slot-all>`, btw? Is it coming later as a separate feature?

**Comment by tomalec** [comment]
Thanks a lot for the review!
> What about the wildcard names `<slot-all name="myapp/*"></slot-all>`, btw? Is it coming later as a separate feature?

Yes, I wanted to start simplest possible, to unlock `uni-menu` work. Patterns will come later as a separate feature

**Comment by tomalec** [comment]
It seems there is another ShadyDOM bug that heavily affects the functionality of the element, as mutation observer is not notified. @platosha have you faced such issue before?

i isolated the issue to https://github.com/webcomponents/shadydom/issues/260

**Comment by platosha** [comment]
@tomalec Indeed can reproduce. I wasn’t aware of this issue, but looks like there is an `observeChildren` method in ShadyDOM just for that: https://github.com/webcomponents/shadydom/blob/7df92c0022eef22f2bc37d44da886cf250678585/tests/observeChildren.html#L116

Non-standard API :-(

**Comment by tomalec** [comment]
Thanks! I was not aware of this `childrenObserver`. Unfortunately it solves half of the problem, as it does not trigger for children removal

**Comment by tomalec** [comment]
I addressed the comments and managed to make it pass in polyfilled browser, but with a limitation, that it will not catch elements removed by `.remove`.

@warpech  do you have anything to add?

**Comment by tomalec** [comment]
I addressed the comments and managed to make it pass in polyfilled browser, but with a limitation, that it will not catch elements removed by `.remove`.

@warpech  do you have anything to add?

File: slot-all.html [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+                nodeList.forEach((element)=>{
+                    if(element.nodeType === Node.ELEMENT_NODE){
+                        // element.assignedSlot no longer works,
+                        // we need to find the formerly assigned slot element manually
+                        const slotName = element.getAttribute('slot');
+                        // if host still has the elements with the same slot name,
+                        // we should keep the slot
+                        const slot = __createdSlots.get(slotName);
+                        // Polyfileld browsers does not have assignedElements
```
Comment (warpech) [inline slot-all.html]: `Polyfileld browsers does not`
↓
`Polyfilled browsers do not`

File: README.md [hunk 5]
```diff
@@ -29,44 +27,31 @@ Or [download as ZIP](https://github.com/Juicy/juicy-element/archive/master.zip).
...
-`onsomething` | Triggers when something happens.
-
-## CSS Custom Properties
-
-Name                          | Description
----                           | ---
-`--juicy-element-laser-color` | Color of the shooted lasers.
+WebComponentsJS Shadow DOM polyfill has some limitations.
+It does ignore `node.remove()` and `.insertAdjacentElement()` calls,
+meaning it cannot be observed or reflected in shadow root. This result in `slot-all` element being unable to react on child nodes been added or removed using this methods. If you are targeting polyfilled environments try using `node.parentNode.removeChild(node)` and `insertBefore`.
```
Comment (warpech) [inline README.md]: `this methods` -> `these methods`

File: index.html [hunk 6]
```diff
@@ -1,21 +1,114 @@
...
+                    Distribute all elements for named slots without knowing their names
+                </h3>
+                <slot-all></slot-all>
+                <slot></slot>
+            </template>
+        </div>
+    </div>
+
+
+    <h4>Looking for a more specs? Check out the <a href="test/">test suite</a>.</h4>
```
Comment (warpech) [inline index.html]: `Looking for a more specs?` -> `Looking for more specs?`

File: index.html [hunk 7]
```diff
@@ -1,21 +1,114 @@
...
+            } else {
+                host.shadowRoot.innerHTML = '';
+            }
+            host.shadowRoot.appendChild(document.importNode(templ.content, true));
+        });
+    }
+    // bound editor to the example
+    document.querySelector('juicy-ace-editor').addEventListener('change', (ev)=>{
+        document.querySelector('[shadow-host]').innerHTML = ev.target.value;
+        // trick Ace editor to adatpt to the new content
```
Comment (warpech) [inline index.html]: `adatpt` -> `adapt`

File: slot-all.html [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                        // following is not supported by polyfill
+                        // slotAll.insertAdjacentElement('beforebegin', namedSlot);
+                        // workaround https://github.com/webcomponents/shadydom/issues/263
+                        slotAll.parentNode.insertBefore(namedSlot, slotAll);
+                    }
+                });
+            }
+            /**
+             * Remove slots for given elements, if needed:
+             *  - there is no more host's children with a slot name for given slot name
```
Comment (warpech) [inline slot-all.html]: `there is no more host's children with a slot name for given slot name`

↓

`there are no more host's children with a slot name for the given slot name`

File: test/spec.html [hunk 9]
```diff
@@ -8,28 +8,161 @@
...
         });
 
         it('should be upgraded element', function() {
-            expect(customElements.get('juicy-element')).to.be.defined;
-            expect(juicyElement).to.be.an.instanceOf(customElements.get('juicy-element'));
+            expect(customElements.get('slot-all')).to.be.defined;
+            expect(host.shadowRoot.querySelector('slot-all')).to.be.an.instanceOf(customElements.get('slot-all'));
+        });
+
+        it('should create `slot` elements for all slot names used by host\'s children, and `all-slotted` attribute to them', function() {
```
Comment (warpech) [inline test/spec.html]: > and `all-slotted`

↓

> and add `all-slotted`

**Review by warpech — APPROVED** [review]
The code looks really good to me, so I decided to do a typo check instead :)

**Review by warpech — COMMENTED** [review]
(ignore this comment)

**Pull request merged**